packages/langium/src/lsp/formatter.ts

```diff
@@ -200,11 +204,9 @@ export abstract class AbstractFormatter implements Formatter {
         return false;
     }
 
-    /**
-     * @deprecated This method has been deprecated with 3.1. It now always returns `true` and is no longer used by the default formatter implementation.
-     */
-    protected isNecessary(_edit: TextEdit, _document: TextDocument): boolean {
-        return true;
+
+    protected isNecessary(edit: TextEdit, document: TextDocument): boolean {
+        return edit.newText !== document.getText(edit.range);
```

Comment:
Suggestion: I've already noticed in https://github.com/eclipse-langium/langium/pull/1550 that this comparison yields faulty behavior on Windows. `document.getText()` will return line endings with `\r`, while `edit.newText` will not contain `\r`.
```suggestion
        return edit.newText !== document.getText(edit.range).replace(/\r/g, '');
```

Reply:
Test case added and fixed `\r` on Windows.